internal/transport/http2_client_test.go

```diff
@@ -0,0 +1,210 @@
+
+	incorrectPreface := "INCORRECT PREFACE\r\n\r\n"
+	n, err = hc.Conn.Write([]byte(incorrectPreface))
+	return n, err
+}
+
+func dialerClientPrefaceLength(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &clientPrefaceLength{Conn: conn}, nil
+}
+
+func (hc *framerWriteSettings) Write(b []byte) (n int, err error) {
+	n, err = hc.Conn.Write(b)
+	//compare framer value
+	if n == 9 {
+		return 0, errors.New("Framer write setting error")
+	}
+	return n, err
+}
+
+func dialerFramerWriteSettings(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &framerWriteSettings{Conn: conn}, nil
+}
+
+func (hc *framerWriteWindowUpdate) Write(b []byte) (n int, err error) {
+
+	n, err = hc.Conn.Write(b)
+	// compare for windowupdate value
+	if n == 13 {
+		return 0, errors.New("Framer write windowupdate error")
+	}
+	return n, err
+}
+
+func dialerFramerWriteWindowUpdate(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &framerWriteWindowUpdate{Conn: conn}, nil
+}
+
+func TestNewHTTP2ClientPrefaceFailure(t *testing.T) {
+
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	defer func() {
+		if r := recover(); r != nil {
+			t.Fatalf("TestNewHTTP2ClientPrefaceFailure panicked: %v", r)
+		}
+	}()
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(2*time.Second))
+	defer cancel()
+	_, err = NewClientTransport(ctx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, ConnectOptions{Dialer: dialerClientPrefaceWrite}, func(GoAwayReason) {})
+	if err == nil {
+		t.Error("Expected an error, but got nil")
+	} else {
+		if err.Error() != "connection error: desc = \"transport: failed to write client preface: preface write error\"" {
+			t.Fatalf("Error while creating client transport: %v", err)
+		}
+	}
+}
+
+func TestNewHTTP2ClientPrefaceLengthFailure(t *testing.T) {
+	// Create a server.
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening: %v", err)
+	}
+	defer lis.Close()
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(2*time.Second))
+	defer cancel()
+	defer func() {
+		if r := recover(); r != nil {
+			t.Fatalf("TestNewHTTP2ClientPrefaceLengthFailure panicked: %v", r)
+		}
+	}()
+
+	_, err = NewClientTransport(ctx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, ConnectOptions{Dialer: dialerClientPrefaceLength}, func(GoAwayReason) {})
+	if err == nil {
+		t.Errorf("Expected an error, but got nil")
+	} else {
+		if err.Error() != "connection error: desc = \"transport: preface mismatch, wrote 21 bytes; want 24\"" {
+			t.Fatalf("Error while creating client transport: %v", err)
```

Comment:
same as line 132. Error message should state got before want

Reply:
done